src/XMakeBuildEngine/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1578,6 +1594,18 @@ private void ReportResultsToSubmission(BuildResult result)
                 if (_buildSubmissions.ContainsKey(result.SubmissionId))
                 {
                     BuildSubmission submission = _buildSubmissions[result.SubmissionId];
+
+                    if (result.OverallResult == BuildResultCode.Success)
+                    {
+                        ILoggingService loggingService = ((IBuildComponentHost)this).LoggingService;
+                        // Check if the user had specified /warnaserror and then override the result if there were any errors.
+                        // This ensures the old behavior stays in tact where builds could succeed even if a failure was logged.
+                        if (loggingService.WarningsAsErrors != null && loggingService.HasBuildSubmissionLoggedErrors(submission.SubmissionId))
+                        {
+                            result.SetOverallResult(overallResult: false);
+                        }
```

Comment:
Is there a good way to abstract this logic? It's pretty repetitive in here. Maybe `EnsureFailureIfWarningsAsErrors(BuildResult, ILoggingService)`?